ui/components/src/TimeRangeSelector/TimeRangeSelector.tsx

```diff
@@ -0,0 +1,113 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { useRef, useState } from 'react';
+import { FormControl, InputLabel, MenuItem, Popover, Select, SelectProps, Stack } from '@mui/material';
+import { format, sub } from 'date-fns';
+import { AbsoluteTimeRange, TimeOption, convertTimeShortcut, parseDurationString } from '@perses-dev/core';
+import { AbsoluteTimePicker } from './AbsoluteTimePicker';
+
+const DATE_TIME_FORMAT = 'yyyy-MM-dd HH:mm:ss';
+const FORM_CONTROL_LABEL = 'Time Range';
+
+interface TimeRangeSelectorProps {
+  timeOptions: TimeOption[];
+  onChange: (value: AbsoluteTimeRange) => void;
+  defaultTimeOption?: TimeOption;
```

Comment:
This part of the API feels a little weird to me for two reasons:

1. I was expecting that the value could be an absolute or relative time range (since the user can select either).
2. I was expecting this to have both `onChange` and a `value` prop so that the state was controlled completely outside this input component.

I don't think we want to have an `onChange` without the component's value also being controlled from outside. Otherwise we're duplicating state, having to sync stuff, etc.

Reply:
Just as an example, basically this is what I was kinda expecting as the API:

```ts
interface TimeRangeSelectorProps {
  timeOptions: TimeOption[];
  value: AbsoluteTimeRange | RelativeTimeRange;
  onChange: (value: AbsoluteTimeRange | RelativeTimeRange) => void;
}
```

Maybe you also end up with something for that union type of relative/absolute too. (e.g. `type TimeRangeValue = AbsoluteTimeRange | RelativeTimeRange`

Reply:
I made progress on this, but left a few TODOs for a follow up PR.